### DISCUSSION TITLE
No serial output on USART1 or USART2

### DISCUSSION THREAD

**Discussion by rberkelm** [SpenceKonde/DxCore#283] [Q&A]
Just testing my new AVR128DB48 Curiosity Nano and I don't seem to get any serial output on USART1 (CDC Rx/CDC Tx pins) or USART2 (PF4/PF5 pins). Am using Tera Term at 9600 baud (8-1-N) and a separate FTDI serial converter. Have tried other combos in Tera Term as well. In my simple blink sketch I am specifying the serial ports as Serial, Serial1 or Serial2. However USART3 (PB0/PB1 pins) work fine, specifying Serial3. Is there anything special in the setup of the first two serial ports?

**Comment by SpenceKonde** [comment]
I think those are used for debugging but not sure - all the debugging baggage is the main reason I dont use curiosity nanos,.

**Reply by SpenceKonde** [reply]
Any luck?

When you're using the alternate pins (pin 4 and 5, are doing swap() before begin() like you're supposed to? are you using the latest version of the core?

**Reply by rberkelm** [reply]
Actually, I didn't know about about the swap(). So I decided to do a little more investigation, adding Wire.swap() before Wire.begin() in setup. I tried addressing the serial ports as Serial, Serial1, Serial2, Serial3 etc and listening for output on each of the Tx pins. The upshot is that Wire.swap(0), Wire.swap(1), Wire.swap(2) made no difference - no output on any Tx pin. However, addressing the serial port as Serial2 or Serial3 did give me output on some of the Tx pins (with or without Wire.swap()). Serial3 gave me output on Pin_PB0 and also the CDCRX pin. Curiously, Serial2 gave me output on PIN_PC0, nothing else. This is somewhat weird as PC0 is NOT a USART pin on the DB48 version of the chip (but it is on the DA48 version). And I definitely have a AVR128DB48 chip on my curiosity nano and I definitely compiled it for the AVR128DB48. My Dx core version is 1.4.10.
Overall I'm not worried about it, as long as I have serial on one of the pins, I'm more than happy. I plan to go with the DA32 version for my boards and will break out all the USART pins, to be sure... :)
Thanks for your awesome work on the Dx core - you are a legend!!!!!!!!!!!!

**Reply by MX682X** [reply]
Are you sure you are using Wire.swap() and not SerialX.swap()? Wire is the object for I2C....

**Reply by SpenceKonde** [reply]
One thing that's worth noting is that a couple of hours ago I found and fixed a bug that looks like it would have totally destroyed serial RX. But TX should be fine.

**Reply by MX682X** [reply]
about that: I think we can elimante one of the global variables in the Stream class and replace it with a local one (fun side fact: the Stream functions with timeout are broken with disabled millis). It should not impact anything since it is in the "protected" space anyway. The problem is that this variable is "above" the UART class variables so it will require an adjustment on the IRQ handlers.... (if I remember correctly, that is)

**Comment by rberkelm** [comment]
> Are you sure you are using Wire.swap() and not SerialX.swap()?

Yep

**Reply by MX682X** [reply]
please try using SerialX.swap() and check if it returns true. Then see if you get any output on the expected pin
